<!doctype html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>Document</title>
  </head>

  <body>
    <button onclick="click2()">触发</button>
  </body>
  <script>
    //异步变同步
    //有多少个接口 按照顺序请求
    //Promise
    //async  await  +fetch

    //手写一个promise
    function click2(e) {
      console.log('qwe click')
      click_handle(123)
    }
    var click_handle
    function fn() {
      var p = new Promise(function (resolve, reject) {
        // setTimeout(() => {

        click_handle = resolve
        // resolve("666")
        // reject("888")
        // }, 2000);
      })
      return p
    }
    // fn().then(function (result) {
    //   alert(result);
    // }, function (result) {
    //   alert(result);
    // })
    // function fn2() {
    //   return 'fn2执行'
    // }

    // var aa = fn().then((val) => {
    //   console.log('qwe 11', val);
    //   return 'asdfasdfasdf'
    //   return fn2()
    // }).then((val) => {
    //   console.log('qwe 22', val);
    // }).then(() => {
    //   console.log('qwe 333',);
    // })
    // aa.then(() => {
    //   console.log('qwe 99999',);
    // })

    // console.log('qwe aa', aa.then(), bb);

    /*function MyPromise() {


  }
  MyPromise.prototype.then = function(fn1, fn2) {
      var num = Math.round(Math.random() * 100);
      if (num > 50) {
          fn1(num + "大于等于50");
      } else {
          fn2(num + "小于50");
      }
  }

  var my = new MyPromise();
  my.then(function(result) {
      alert("成功的结果:" + result);
  }, function(result) {
      alert("失败的结果:" + result);
  })*/
  </script>
  <script>
    var fn3 = function fn3(params) {
      let p = new Promise((resolve, reject) => {
        setTimeout(() => {
          // console.log('qwe ', qwe);
          console.log('qwe 000')
          resolve(99)
          // reject('123')
        }, 5000)
      })
      // .catch(err => {
      //   console.log('错误！！！')
      // })
      return p
    }
    async function fn4() {
      try {
        console.log('qwe 11111111')
        let aa = await fn3()
        console.log('qwe ', 222)
      } catch (error) {
        console.error(error)
        console.log('qwe!!!!! ')
      }
      console.log('qwe 666')
    }
    // fn4()
    console.log('qwe ', 999)

    async function run() {
      try {
        console.log('qwe jjj')
        // await Promise.reject(new Error('qwe'))
        // await fn3()
        console.log('qwe jjj 222', await fn3())
      } catch (error) {
        console.log(error)
      }
    }
    run()
    console.log('qwe lll')
  </script>
</html>
